src/examples/libpmem2/unsafe_shutdowns/unsafe_shutdowns.c

```diff
@@ -0,0 +1,607 @@
+
+	*file_in_use = FILE_ARMED;
+	pool->persist(file_in_use, sizeof(*file_in_use));
+	int ret = pmem2_deep_sync(pool->map, file_in_use, sizeof(*file_in_use));
+	if (ret) {
+		pmem2_perror("pmem2_deep_sync file in use failed");
+		return ret;
+	}
+
+	return 0;
+}
+
+/*
+ * pool_disarm -- indicate pool modifications are completed
+ */
+static int
+pool_disarm(struct pool_data *pool)
+{
+	uint8_t *file_in_use = &pool->content->ps.file_in_use;
+
+	assert(*file_in_use == FILE_ARMED);
+
+	/* deep sync whole mapping to make sure all persists are completed */
+	pmem2_deep_sync(pool->map, pool->content, pool->map_size);
+
+	*file_in_use = FILE_UNARMED;
+	pool->persist(file_in_use, sizeof(*file_in_use));
+	int ret = pmem2_deep_sync(pool->map, file_in_use, sizeof(*file_in_use));
+	if (ret) {
+		pmem2_perror("pmem2_deep_sync file in use failed");
+		return ret;
+	}
+
+	return 0;
+}
+
+#define USAGE_STR "usage: %s <command> <file> [<arg>]\n" \
+	"Where available commands are:\n" \
+	"\tinit - initialize the file metadata\n" \
+	"\treset - zero file contents and reinit file metadata\n" \
+	"\tread - print file contents\n" \
+	"\twrite - store <arg> into the file\n"
+
+
+int
+main(int argc, char *argv[])
+{
+	struct pmem2_config *cfg = NULL;
+	struct pool_data pool = {0};
+
+	/* parse and validate arguments */
+	if (argc < 3) {
+		fprintf(stderr, USAGE_STR, argv[0]);
+		return 1;
+	}
+
+	int mode_init = (strcmp(argv[1], "init") == 0);
+	int mode_reset = (strcmp(argv[1], "reset") == 0);
+	int mode_read = (strcmp(argv[1], "read") == 0);
+	int mode_write = (strcmp(argv[1], "write") == 0);
+	char *file = argv[2];
+	char *str = NULL;
+
+	if (mode_init + mode_reset + mode_read + mode_write == 0) {
+		/* either create, read or write mode has to be chosen */
+		fprintf(stderr, USAGE_STR, argv[0]);
+		return 1;
+	}
+
+	if (mode_write) {
+		if (argc < 4) {
+			fprintf(stderr, USAGE_STR, argv[0]);
+			return 1;
+		}
+		str = argv[3];
+	}
+
+	/* open file and prepare source */
+	int fd = open(file, O_RDWR);
+	if (fd < 0) {
+		perror(file);
+		return 1;
+	}
+
+	int ret = pmem2_source_from_fd(&pool.src, fd);
+	if (ret) {
+		pmem2_perror("pmem2_source_from_fd");
+		goto err_source;
+	}
+
+	/* prepare configuration */
+	ret = pmem2_config_new(&cfg);
+	if (ret) {
+		pmem2_perror("pmem2_config_new");
+		goto err_config;
+	}
+
+	ret = pmem2_config_set_required_store_granularity(cfg,
+			PMEM2_GRANULARITY_PAGE);
+	if (ret) {
+		pmem2_perror("pmem2_config_set_required_store_granularity");
+		goto err_granularity;
+	}
+
+	/* prepare the mapping */
+	ret = pmem2_map(cfg, pool.src, &pool.map);
+	if (ret) {
+		pmem2_perror("pmem2_map");
+		goto err_map;
+	}
+
+	pool.content = pmem2_map_get_address(pool.map);
+	pool.usable_space_size =
+		pmem2_map_get_size(pool.map) - sizeof(struct pool_state);
+	pool.persist = pmem2_get_persist_fn(pool.map);
+	pool.memset = pmem2_get_memset_fn(pool.map);
+	pool.memcpy = pmem2_get_memcpy_fn(pool.map);
+
+	if (mode_init) {
```

Comment:
why the modes?

This should be demonstrating how an app would actually do things. Which in my opinion should do:

```
if ((state = device_state_verify() == UNINITIALIZED)) {
  device_state_init();
} else if (state == CORRUPTED) {
  fail();
}

arm();

/* normal app things */
if (string is set)
 print(string);
else
  string = hello world;

disarm();
```

A realistic application would not have two separate "create" and "read" modes.